model/CSVDialog.cc

```diff
@@ -34,6 +53,111 @@ void CSVDialog::reportFromFile(const std::string& input, const std::string& outp
   reportFromCSVFile(is,of,spec);
 }
 
+// Return file name after downloading a CSV file from the web.
+std::string CSVDialog::loadWebFile(const std::string& url)
+{
+  // Parse input URL. Also handles URLs of the type username:password@example.com/pathname#section. See https://stackoverflow.com/questions/2616011/easy-way-to-parse-a-url-in-c-cross-platform
+  boost::regex ex(R"((http|https)://([^/ :]+):?([^/ ]*)(/?[^ #?]*)\\??([^ #]*)#?([^ ]*)|^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?)");
+  boost::cmatch what;
+  if (regex_match(url.c_str(), what, ex)) {
+   // what[0] contains the whole string 	 
+   // what[1] is the protocol
+   // what[2] is the domain  
+   // what[3] is the port    
+   // what[4] is the path    
+   // what[5] is the query   
+   // what[6] is the fragment		  
+  } else throw runtime_error("Failure to match URL: "+url);
+         
+  auto const protocol =what[1];
+  auto const host = what[2];
+  auto const port = what[3];
+  auto const target = what[4];
+  auto const query = what[5];
+  auto const fragment = what[6];  
+  
+  // The io_context is required for all I/O
+  boost::asio::io_context ioc;
+  
+  // The SSL context is required, and holds certificates
+  ssl::context ctx{ssl::context::tls_client};
+  
+  // Verify the remote server's certificate. See https://github.com/djarek/certify/blob/master/examples/get_page.cpp
+  ctx.set_verify_mode(ssl::verify_peer | ssl::context::verify_fail_if_no_peer_cert);    
+  ctx.set_default_verify_paths();
+  
+  // tag::ctx_setup_source[]
+  boost::certify::enable_native_https_server_verification(ctx);
+  // end::ctx_setup_source[]        
+  		        
+  // These objects perform our I/O
+  tcp::resolver resolver{ioc};
+  ssl::stream<tcp::socket> stream{ioc, ctx};        
+  
+  // tag::stream_setup_source[]. See https://github.com/djarek/certify/blob/master/examples/get_page.cpp
+  boost::certify::set_server_hostname(stream, host.str());
+  boost::certify::sni_hostname(stream, host);
+  // end::stream_setup_source[]         
+
+  // Look up the domain name
+  auto const results = resolver.resolve(host, protocol);
```

Comment:
Why is the protocol required to resolve the host?

Reply:
One could use either the protocol or the port, I've seen examples doing both online.